src/dachshund/cnm_communities.rs

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+extern crate nalgebra as na;
+use crate::dachshund::graph_base::GraphBase;
+use crate::dachshund::id_types::NodeId;
+use crate::dachshund::node::Node;
+use na::{DMatrix, DVector};
+use ordered_float::OrderedFloat;
+use rand::distributions::WeightedIndex;
+use rand::prelude::*;
+use rand::seq::SliceRandom;
+use rand::Rng;
+use std::cmp::Ordering;
+use std::collections::{BTreeSet, BinaryHeap, HashMap, HashSet, VecDeque};
+use std::collections::hash_map::Keys;
+use std::iter::FromIterator;
+type Community = HashSet<NodeId>;
+
+#[derive(Clone, Copy, Eq)]
+pub struct CNMCommunityMergeInstruction {
+    delta_ij: OrderedFloat<f64>,
+    i: usize,
+    j: usize,
+}
+impl CNMCommunityMergeInstruction {
+    pub fn new(delta_ij: OrderedFloat<f64>, i: usize, j: usize) -> Self {
+        Self { delta_ij, i, j }
+    }
+    pub fn tuple(self) -> (OrderedFloat<f64>, usize, usize) {
+        (self.delta_ij, self.i, self.j)
+    }
+}
+impl Ord for CNMCommunityMergeInstruction {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.delta_ij < other.delta_ij {
+            Ordering::Less
+        } else if self.delta_ij > other.delta_ij {
+            Ordering::Greater
+        } else if self.i > other.i {
+            Ordering::Less
+        } else if self.i < other.i {
+            Ordering::Greater
+        } else if self.j > other.j {
+            Ordering::Less
+        } else if self.j < other.j {
+            Ordering::Greater
+        } else {
+            Ordering::Equal
+        }
+    }
+}
+impl PartialEq for CNMCommunityMergeInstruction {
+    fn eq(&self, other: &Self) -> bool {
+        self.delta_ij == other.delta_ij && self.i == other.i && self.j == other.j
+    }
+}
+impl PartialOrd for CNMCommunityMergeInstruction {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+type CNMCommunityMergeInstructionHeap = BinaryHeap<CNMCommunityMergeInstruction>;
+
+pub trait CNMCommunities : GraphBase {
+    fn get_max_maxheap(
+        &self,
+        delta_q_maxheap: &HashMap<usize, CNMCommunityMergeInstructionHeap>,
+    ) -> CNMCommunityMergeInstructionHeap {
+        let mut maxh: CNMCommunityMergeInstructionHeap = BinaryHeap::new();
+        for (_k, heap) in delta_q_maxheap.iter() {
+            let maybe_top_elem = heap.peek();
+            if maybe_top_elem.is_some() {
+                let top_elem = maybe_top_elem.unwrap();
+                maxh.push(top_elem.clone());
+            }
+        }
+        maxh
+    }
+    fn init_cnm_communities(
+        &self,
+    ) -> (
+        HashMap<usize, Community>,
```

Comment:
Can you include comments explaining the semantics of each of the elements of the returned tuple?

Reply:
I am thinking of adding a private struct called CNMIntermediaryState, or something like that, so we can just have self-documenting fields.